File: 2015/10/rs/src/main.rs
fn get_next_value(value: &str) -> String {
    let mut result = String::new();
    let mut last_digit = '\0';
    let mut length = 0;
    for c in value.chars() {
        if c == last_digit {
            length += 1;
        } else {
            if last_digit != '\0' {
                result.push_str(length.to_string().as_str());
                result.push(last_digit);
            }
            last_digit = c;
            length = 1;
        }
    }
    result.push_str(length.to_string().as_str());
    result.push(last_digit);
    result
}

fn solve(puzzle_input: &str) -> (usize, usize) {
    let mut current_value = String::from(puzzle_input);
    let mut part1 = 0;
    for turn in 0..50 {
        if turn == 40 {
            part1 = current_value.len();
        }
        current_value = get_next_value(&current_value);
    }
    (part1, current_value.len())
}

fn get_input(file_path: &String) -> String {
    String::from(
        std::fs::read_to_string(file_path)
            .expect("Error reading input file!")
            .trim(),
    )
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2015/10/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>
#include <math.h>

typedef char *Input;
typedef struct
{
    int part1;
    int part2;
} Results;

typedef struct {
    char c;
    int length;
} Sequence;

typedef struct {
    Sequence *sequences;
    int size, capacity;
} Result;
#define SEQUENCE_INCREMENT 100

void addToResult(Result *result, int length, char c)
{
    if (result->size == result->capacity)
        result->sequences = realloc(result->sequences, (result->capacity += SEQUENCE_INCREMENT) * sizeof(Sequence));
    result->sequences[result->size++] = (Sequence){ c, length };
}

char *processResult(Result result)
{
    char *sequence = calloc((result.size * 2), sizeof(char));
    char *start = sequence;
    for (int index = 0; index < result.size; index++)
    {
        Sequence current = result.sequences[index];
        *sequence = current.length + '0';
        sequence++;
        *sequence = current.c;
        sequence++;
    }
    free(result.sequences);
    return start;
}

char *getNextValue(char *value)
{
    Result result = {
        malloc(SEQUENCE_INCREMENT * sizeof(Sequence)),
        0, SEQUENCE_INCREMENT
    };
    unsigned int valueLength = strlen(value);
    int length = 1;
    char lastDigit = value[0];
    for (unsigned int index = 1; index < valueLength; index++)
    {
        if (value[index] == lastDigit)
            length++;
        else
        {
            addToResult(&result, length, lastDigit);
            lastDigit = value[index];
            length = 1;
        }
    }
    addToResult(&result, length, lastDigit);
    free(value);
    return processResult(result);
}

Results solve(Input input)
{
    char *currentValue = input;
    int part1 = 0;
    for (int turn = 0; turn < 50; turn++)
    {
        if (turn == 40)
            part1 = strlen(currentValue);
        currentValue = getNextValue(currentValue);
    }
    return (Results){part1, strlen(currentValue)};
}

char *rtrim(char *str, const char *seps)
{
    int i;
    if (seps == NULL)
        seps = "\t\n\v\f\r ";
    i = strlen(str) - 1;
    while (i >= 0 && strchr(seps, str[i]) != NULL)
        str[i--] = '\0';
    return str;
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    fclose(file);
    return rtrim(content, NULL);
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
